### PR TITLE
Initial change for v2.

### PR DESCRIPTION
**Pull request by bwplotka**
As per https://github.com/grpc-ecosystem/go-grpc-middleware/issues/275

The main goals are to have separate modules per each implementation and all reusing the same code.
This way we can ensure consistency, and we reduce LOC to a minimum, including tests.

## Changes:

* [x] Removed grpc_ prefix from all packages like `grpc_middleware`. Users can add them on their own.
* [x] Moved all specific implementations to separate modules. The exception is opentracing, as opentracing itself
is just an amazing interface, so.. perfect (:
* [x] Added generic interceptor code that both monitoring, logging, and tracing will use under /interceptors
* [x] Reuse generic interceptors in interceptor/{logging,tracing,metrics,tags}. This allows huge simplification of those.
* [x] No context logger. Only ctxtags.Tags responsible for propagating tags in context and proto messages.
* [x] Tags are now map[string]string to discourage context value abuse.
* [x] PayloadUnary works as a standalone interceptor as well.
* [x] Tags will actually use map tags instead of noop if not created.
* [x] Added scripts for a proto generation.
* [x] Let tag.Interceptor to extract request data for all types of requests.

## New package structure:

```
├── auth
│   └── [util package for auth for gRPC]
├── chain.go
├── chain_test.go
├── CHANGELOG.md
├── CONTRIBUTING.md
├── doc.go
├── go.mod
├── go.sum
├── grpctesting
│   └── [util package for testing and test suite]
├── interceptors
│   ├── client.go [core package for all interceptors]
│   ├── client_test.go
│   ├── doc.go
│   ├── logging
│   │    └── [logging interceptor package allowing to inject Logger]
│   ├── ratelimit
│   │   └── [ratelimit interceptor]
│   ├── recovery
│   │   └── [panic recovery interceptor]
│   ├── reporter.go
│   ├── retry
│   │   └── [retry interceptor]
│   ├── server.go
│   ├── server_test.go
│   ├── tags
│   │   └── [interceptor allowing to inject and propagate string tags]
│   ├── tracing
│   │   └── [opentracing interceptor]
│   └── validator
│       └── [proto validation interceptor]
├── LICENSE
├── Makefile
├── providers [All implementations for our interfaces for each provider. Always separate modules]
│   ├── kit
│   │   ├── examples_test.go
│   │   ├── go.mod
│   │   ├── go.sum
│   │   └── logger.go
│   ├── logrus
│   │   ├── examples_test.go
│   │   ├── go.mod
│   │   ├── go.sum
│   │   └── logger.go
│   └── zap
│       ├── examples_test.go
│       ├── go.mod
│       ├── go.sum
│       └── logger.go
├── README.md
├── scripts
├── util
│   └── [utils]
├── wrappers.go
└── wrappers_test.go

21 directories, 98 files
```

Signed-off-by: Bartlomiej Plotka <bwplotka@gmail.com>

### CODE REVIEW

**Comment by bwplotka** [comment]
Hi, I know it's a huge change, but I believe it's ready to go.

Missing unit test for zap,logrus,kit, but will add later. The implementation are ridiculously small now. 

Happy to schedule some call with you @domgreen @johanbrandhorst to go through this! :muscle:

File: grpctesting/testproto/test.manual_extractfields.pb.go [hunk 1]
```diff
@@ -0,0 +1,10 @@
+// Manual code for logging field extraction tests.
+
+package grpc_middleware_testproto
```
Comment (johanbrandhorst) [inline grpctesting/testproto/test.manual_extractfields.pb.go]: Can we make this `testproto`?

File: grpctesting/testproto/test.manual_validator.pb.go [hunk 2]
```diff
@@ -1,11 +1,11 @@
 // Manual code for validation tests.
 
-package mwitkow_testproto
+package grpc_middleware_testproto
```
Comment (johanbrandhorst) [inline grpctesting/testproto/test.manual_validator.pb.go]: `testproto`?

File: interceptors/validator/validator.go [hunk 3]
```diff
@@ -1,7 +1,7 @@
 // Copyright 2016 Michal Witkowski. All Rights Reserved.
 // See LICENSE for licensing terms.
 
-package grpc_validator
+package grpcvalidator
```
Comment (johanbrandhorst) [inline interceptors/validator/validator.go]: Do we need the `grpc` prefix?

File: providers/kit/go.mod [hunk 4]
```diff
@@ -0,0 +1,11 @@
+module github.com/grpc-ecosystem/go-grpc-middleware/providers/kit
+
+go 1.13
+
+require (
+	github.com/go-kit/kit v0.10.0
+	github.com/grpc-ecosystem/go-grpc-middleware v1.2.0
```
Comment (johanbrandhorst) [inline providers/kit/go.mod]: Should this use v2?

File: auth/auth.go [hunk 5]
```diff
@@ -1,12 +1,12 @@
...
 // See LICENSE for licensing terms.
 
-package grpc_auth
+package auth
 
 import (
 	"context"
 
-	"github.com/grpc-ecosystem/go-grpc-middleware"
+	middleware "github.com/grpc-ecosystem/go-grpc-middleware"
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply auth/auth.go]: yea, as it is different than folder name ):

File: auth/examples_test.go [hunk 6]
```diff
@@ -1,16 +1,16 @@
-package grpc_auth_test
+package auth_test
 
 import (
 	"context"
 
-	"github.com/grpc-ecosystem/go-grpc-middleware/auth"
-	"github.com/grpc-ecosystem/go-grpc-middleware/tags"
+	grpcauth "github.com/grpc-ecosystem/go-grpc-middleware/auth"
+	grpc_ctxtags "github.com/grpc-ecosystem/go-grpc-middleware/interceptors/tags"
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply auth/examples_test.go]: Sorry leftovers will fix

File: grpctesting/interceptor_suite.go [hunk 7]
```diff
@@ -1,7 +1,7 @@
 // Copyright 2016 Michal Witkowski. All Rights Reserved.
 // See LICENSE for licensing terms.
 
-package grpc_testing
+package grpctesting
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply grpctesting/interceptor_suite.go]: We don't but in the same time we use this in unit test so `testing` package is *always* there. Also it's could be internal package I think

File: interceptors/logging/payload.go [hunk 8]
```diff
@@ -0,0 +1,155 @@
+package logging
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/golang/protobuf/jsonpb"
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply interceptors/logging/payload.go]: yes, but maybe in following PR?

File: interceptors/recovery/examples_test.go [hunk 9]
```diff
@@ -22,17 +22,17 @@ func Example_initialization() {
...
-		grpc_recovery.WithRecoveryHandler(customFunc),
+	opts := []recovery.Option{
+		recovery.WithRecoveryHandler(customFunc),
 	}
 	// Create a server. Recovery handlers should typically be last in the chain so that other middleware
 	// (e.g. logging) can operate on the recovered state instead of being directly affected by any panic
 	_ = grpc.NewServer(
-		grpc_middleware.WithUnaryServerChain(
-			grpc_recovery.UnaryServerInterceptor(opts...),
+		middleware.WithUnaryServerChain(
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply interceptors/recovery/examples_test.go]: Oh, will check

File: interceptors/ratelimit/examples_test.go [hunk 10]
```diff
@@ -1,8 +1,8 @@
 package ratelimit_test
 
 import (
-	"github.com/grpc-ecosystem/go-grpc-middleware"
-	"github.com/grpc-ecosystem/go-grpc-middleware/ratelimit"
+	middleware "github.com/grpc-ecosystem/go-grpc-middleware"
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply interceptors/ratelimit/examples_test.go]: Cannot, go imports adds it

File: interceptors/retry/retry_test.go [hunk 11]
```diff
@@ -11,9 +11,9 @@ import (
 	"time"
 
 	grpc_middleware "github.com/grpc-ecosystem/go-grpc-middleware"
-	grpc_retry "github.com/grpc-ecosystem/go-grpc-middleware/retry"
-	"github.com/grpc-ecosystem/go-grpc-middleware/testing"
-	pb_testproto "github.com/grpc-ecosystem/go-grpc-middleware/testing/testproto"
+	grpc_testing "github.com/grpc-ecosystem/go-grpc-middleware/grpctesting"
+	pb_testproto "github.com/grpc-ecosystem/go-grpc-middleware/grpctesting/testproto"
+	grpc_retry "github.com/grpc-ecosystem/go-grpc-middleware/interceptors/retry"
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply interceptors/retry/retry_test.go]: sorry will fix

File: interceptors/validator/validator.go (hunk 3 above)
Comment (bwplotka) [reply interceptors/validator/validator.go]: here probably not

File: interceptors/logging/payload.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (bwplotka) [reply interceptors/logging/payload.go]: added TODO: https://github.com/grpc-ecosystem/go-grpc-middleware/issues/275

File: interceptors/recovery/examples_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (bwplotka) [reply interceptors/recovery/examples_test.go]: We would need to upgrade gRPC. Added TODO: https://github.com/grpc-ecosystem/go-grpc-middleware/issues/275

File: CHANGELOG.md [hunk 12]
```diff
@@ -25,7 +25,7 @@ Types of changes:
 - [#181](https://github.com/grpc-ecosystem/go-grpc-middleware/pull/181) Rate Limit support - [ceshihao](https://github.com/ceshihao)
 - [#161](https://github.com/grpc-ecosystem/go-grpc-middleware/pull/161) Retry on server stream call - [lonnblad](https://github.com/lonnblad)
 - [#152](https://github.com/grpc-ecosystem/go-grpc-middleware/pull/152) Exponential backoff functions - [polyfloyd](https://github.com/polyfloyd)
-- [#147](https://github.com/grpc-ecosystem/go-grpc-middleware/pull/147) Jaeger support for ctxtags extraction - [vporoshok](https://github.com/vporoshok)
+- [#147](https://github.com/grpc-ecosystem/go-grpc-middleware/pull/147) Jaeger support for tags extraction - [vporoshok](https://github.com/vporoshok)
```
Comment (johanbrandhorst) [inline CHANGELOG.md]: I think this was probably a mistake?

Comment (bwplotka) [reply CHANGELOG.md]: yes, but also does not matter, does it?

**Pull request merged**